### PR TITLE
Staff Panel Account List

### PR DESCRIPTION
**Pull request by freddyheppell**
## Overview of Changes

- [x] add account list
- [x] link to forum admin panel
- activation
    - [x] show if user is activated
    - [x] manually activate (probably no need to deactivate?)
    - [x] resend activation email
- email changes
    - [x] list email changes
    - [x] manually approve email change
- groups 
     - [x] show groups
     - [x] set new user groups
     - [x] remove a group
     - [x] ~~prevent staff setting higher than their own group~~ (or, only let admins do it?)
- donations
    - [x] view donation list
    - [ ] _edit a donation_
    - [ ] _create a donation_
- game accounts
    - [x] list game accounts
    - [x] remove game account
    - [ ] manual add (?)
    - [x] ban status
- [x] edit users
- [ ] audit log (?)
- [ ] delete user (?)
- [x] prevent non-staff from viewing
- [x] forum sso sync
- [x] force SSO sync
- [x] enable group syncing

_italics_ = linked on the users page somehow but not actually part of it, will implement later

## Changes made on deployment
* Fill in discourse_name column in db for ranks

### CODE REVIEW

**Comment by freddyheppell** [comment]
Having accounts and unactivated accounts on separate tables will make it a bit difficult to include them on this list. Is there a particular reason for that which I'm missing instead of just having one users table with a boolean `activated` column?

**Comment by andyksaw** [comment]
Now that I think about it, there's not really any benefit having unactivated accounts in a separate table.

My initial thought was that you could register with the same email address unlimited times provided that there's no activated account with that email - but that's totally doable anyway :thinking:

**Comment by freddyheppell** [comment]
> My initial thought was that you could register with the same email address unlimited times provided that there's no activated account with that email

That's not actully possible any more... that was an edge case I realised when writing registration tests and patched that out...

If you want I could refactor on the DB layer to put all this on one table. I'll keep it so there's still two separate repositories but they just resolve to the same table with the relevant acoccount types filtered.

**Comment by andyksaw** [comment]
> If you want I could refactor on the DB layer to put all this on one table. I'll keep it so there's still two separate repositories but they just resolve to the same table with the relevant acoccount types filtered.

Sounds good 👍

**Comment by freddyheppell** [comment]
There's a couple of features I'm not sure if we want or not, any thoughts? Also I need to think up which ranks get access to this or not. We already have staff vs admin in the DB (although SOPs aren't classified as admins which we might want to change) so we could just do it by that.

**Comment by freddyheppell** [comment]
I'm looking at group syncing and we need some way of being able to give discourse a comma separated list of group _names_ rather than IDs as we currently use to manually add one group (when people donate). What do you think would be the best way to do this as we currently both have groups in the DB and [in code](https://github.com/andyksaw/ProjectCityBuild/blob/68006b263006de7488a8ecfb6fe39eadeecfc27a/src/app/Entities/Groups/GroupEnum.php)

**Comment by freddyheppell** [comment]
I believe all the core features are now in this PR, I'd like to write some tests before merging though.

File: src/app/Http/Controllers/DonationController.php [hunk 1]
```diff
@@ -99,15 +91,11 @@ public function donate(Request $request)
 
         // add user to donator group if they're logged in
         if ($account !== null) {
-            $group = new GroupEnum(GroupEnum::Donator);
-            $donatorGroup = $this->groupRepository->getGroupByName(GroupEnum::Donator);
+            $donatorGroup = $this->groupRepository->getGroupByName("donator");
```
Comment (andyksaw) [inline src/app/Http/Controllers/DonationController.php]: I wonder if there's a better way than specifying "donator" as a string here 🤔

File: src/app/Http/Controllers/Panel/AccountController.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+     * @return \Illuminate\Http\Response
+     */
+    public function update(Request $request, Account $account)
+    {
+        $account->update($request->all());
+        $account->save();
+
+        $account->emailChangeRequests()->delete();
+
+        $syncAction = resolve(SyncUserToDiscourse::class);
```
Comment (andyksaw) [inline src/app/Http/Controllers/Panel/AccountController.php]: Can this not be injected into the method?
```php
public function update(Request $request, Account $account, SyncUserToDiscourse $syncAction) {
```

File: src/app/Http/Controllers/Panel/AccountDiscourseAdminRedirect.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        $this->discourseUserApi = $discourseUserApi;
+    }
+
+    public function __invoke(Account $account)
+    {
+        $discourseUser = $this->discourseUserApi->fetchUserByPcbId($account->account_id);
+        $discourseId = $discourseUser["user"]["id"];
+        $discourseUsername = $discourseUser["user"]["username"];
+
+        return redirect("https://forums.projectcitybuild.com/admin/users/$discourseId/$discourseUsername");
```
Comment (andyksaw) [inline src/app/Http/Controllers/Panel/AccountDiscourseAdminRedirect.php]: We should probably move the `https://forums.projectcitybuild.com/` part to a reusable constant somewhere

File: src/app/Http/Controllers/Panel/AccountController.php (hunk 2 above)
Comment (freddyheppell) [reply src/app/Http/Controllers/Panel/AccountController.php]: I believe this screws with route model binding, can’t entirely remember though. There definitely was an issue because I didn’t know about the ‘resolve’ helper before this PR so I must’ve looked it up for some reason.

File: src/app/Http/Controllers/Panel/AccountDiscourseAdminRedirect.php (hunk 3 above)
Comment (freddyheppell) [reply src/app/Http/Controllers/Panel/AccountDiscourseAdminRedirect.php]: Possibly yeah, although the chances of us changing the forum URL is near zero and it’d be a fairly simple job if we did want to.

File: src/app/Http/Controllers/DonationController.php (hunk 1 above)
Comment (freddyheppell) [reply src/app/Http/Controllers/DonationController.php]: We could set up a configuration thing that lets us change previously hardcoded things like this
